openfl-tutorials/interactive_api/Director_Pytorch_TinyImageNet/envoy/tinyimagenet_shard_descriptor.py

```diff
@@ -0,0 +1,121 @@
+        self.data_type = data_type
+        self._common_data_folder = data_folder
+        self._data_folder = os.path.join(data_folder, data_type)
+        self.labels = {}  # fname - label number mapping
+        self.image_paths = sorted(
+            glob.iglob(
+                os.path.join(self._data_folder, '**', '*.JPEG'),
+                recursive=True
+            )
+        )[rank - 1::worldsize]
+        with open(os.path.join(self._common_data_folder, 'wnids.txt'), 'r') as fp:
+            self.label_texts = sorted([text.strip() for text in fp.readlines()])
+        self.label_text_to_number = {text: i for i, text in enumerate(self.label_texts)}
+        self.fill_labels()
+
+    def __len__(self) -> int:
+        """Return the len of the dataset."""
+        return len(self.image_paths)
+
+    def __getitem__(self, index: int) -> Tuple['Image', int]:
+        """Return an item by the index."""
+        file_path = self.image_paths[index]
+        label = self.labels[os.path.basename(file_path)]
+        return self.read_image(file_path), label
+
+    def read_image(self, path: Path) -> Image:
+        """Read the image."""
+        img = Image.open(path)
+        return img
+
+    def fill_labels(self) -> None:
+        """Fill labels."""
+        if self.data_type == 'train':
+            for label_text, i in self.label_text_to_number.items():
+                for cnt in range(self.NUM_IMAGES_PER_CLASS):
+                    self.labels[f'{label_text}_{cnt}.JPEG'] = i
+        elif self.data_type == 'val':
+            with open(os.path.join(self._data_folder, 'val_annotations.txt'), 'r') as fp:
+                for line in fp.readlines():
+                    terms = line.split('\t')
+                    file_name, label_text = terms[0], terms[1]
+                    self.labels[file_name] = self.label_text_to_number[label_text]
+
+
+class TinyImageNetShardDescriptor(ShardDescriptor):
+    """Shard descriptor class."""
+
+    def __init__(
+            self,
+            data_folder: str = 'data',
+            rank_worldsize: str = '1,1',
+            **kwargs
+    ):
+        """Initialize TinyImageNetShardDescriptor."""
+        self.common_data_folder = Path.cwd() / data_folder
+        self.data_folder = Path.cwd() / data_folder / 'tiny-imagenet-200'
+        self.download_data()
+        self.rank, self.worldsize = tuple(int(num) for num in rank_worldsize.split(','))
+
+    def download_data(self):
+        """Download prepared dataset."""
+        zip_file_path = self.common_data_folder / 'tiny-imagenet-200.zip'
+        os.makedirs(self.common_data_folder, exist_ok=True)
+        os.system(f'wget --no-clobber http://cs231n.stanford.edu/tiny-imagenet-200.zip'
+                  f' -O {zip_file_path}')
+        shutil.unpack_archive(str(zip_file_path), str(self.common_data_folder))
+
+    def get_dataset(self, dataset_type):
+        """Return a dataset by type."""
+        return TinyImageNetDataset(
```

Comment:
Several questions araise here.
How should the frontend user know that `get_dataset` should be called? <br>
How can we assert the TinyImageNetDataset instance would return samples of the same shape that is stated by the 'sample_shape' property of the SD? <br>

Reply:
The same way as for another implementations and interfaces. It is the interface of ShardDescriptor and user can use it. And it should be implemented in the DummyShardDescriptor when that ShardDescriptor will move to core package. 
It's not harder to find out, that you can call `self.shard_descriptor.get_dataset('train')` than about 
```
train_sd = deepcopy(self.shard_descriptor)
train_sd.set_mode('train')
```
It is possible to add checking of sample shape in `get_dataset` method or otherwise, the same as with `set_mode`.

Reply:
what do you mean by 'ShardDescriptor to the core package'?

Reply:
>It's not harder to find out, that you can call self.shard_descriptor.get_dataset('train') than about
Nevertheless, how should the frontend user know that get_dataset should be called?